// pages/create-article.js
import React, { useState, useEffect, useContext } from 'react';
import { Redirect } from 'react-router-dom';

import useFetch from 'hooks/use-fetch';
import ArticleForm from 'components/article-form';
import { CurrentUserContext } from 'contexts/current-user';

const CreateArticle = () => {
  const [{isLogged}] = useContext(CurrentUserContext);
  const [{response, error}, doFecth] = useFetch('/articles');
  const [isSuccessSubmit, setIsSuccessSubmit] = useState(false);
  const initialValues = {
    title: '',
    description: '',
    body: '',
    tagList: []
  };

  const handleSubmit = (article) => {
    doFecth({
      method: 'POST',
      data: {
        article
      }
    });
  };

  useEffect(() => {
    if (!response) return;

    setIsSuccessSubmit(true);
  }, [response]);

  if (!isLogged) {
    return <Redirect to='/' />
  }

  if (isSuccessSubmit) {
    return <Redirect to={`/articles/${response.article.slug}`} />
  }

  return (
    <ArticleForm 
      onSubmit={handleSubmit} 
      errors={(error && error.errors) || {}} 
      initialValues={initialValues} 
    />
  )
}

export default CreateArticle;
